### PR TITLE
perf(bytecode): fast-path numeric element reads

### LINKED ISSUE #538: perf(bytecode): Op::GetElement lacks the tree-walker's numeric-index fast path
## Description

The tree-walker's computed member read has a numeric-index fast path; the
bytecode VM's `Op::GetElement` does not, so a compiled body's typed-array
element reads take a longer path than the same reads in the tree-walker.

`eval_member` (`src/interpreter/eval/access.rs:745`) checks for a Number key on
a typed array or array object *before* `ToPropertyKey` and calls
`typed_array_get_index` directly — no allocation.

`member_get_computed` (`src/interpreter/bytecode/vm.rs:110`) has no such check.
It calls `to_property_key` unconditionally, which for an integer index does
`(trunc as u32).to_string()` and then `JsPropertyKey::from(String)` — two heap
allocations — and then `get_object_property` re-derives the index from the
string it just built via `canonical_numeric_index_string`. `Op::GetElement` also
re-roots the entire operand stack per access (`root_operand_stack`).

`Op::SetElement`/`member_set_computed` has the same shape, but there the
tree-walker pays the round trip too, so that half is a shared cost rather than a
VM-only asymmetry — still worth a fast path, lower priority than the read.

## How big is it? Bounded, and smaller than it looks

**Correction to this issue's first revision, which claimed the `arith`-vs-`elem`
gap in `bench_opmix.js` measured this cost. It does not.** From the #526 audit
(`docs/perf/2026-08-26/mandreel-bytecode-work-share.md`), with opcode counts
from the `perf-counters` build:

| variant | opcodes / iter | default µs/iter | `--bytecode` µs/iter | saving |
| --- | --- | --- | --- | --- |
| `arith` | 49 | 4.11 | 2.94 | **1.169 µs** |
| `elem` | 55 | 6.36 | 5.17 | **1.193 µs** |

`elem` does two typed-array reads per iteration through the slow path. If that
path cost meaningfully more than the tree-walker's, `elem`'s **absolute** saving
would sit below `arith`'s. It is marginally above. So the penalty is bounded at
well under ~25 ns per read at this resolution — the smaller `elem` percentage
(-18.8% vs -28.4%) comes from element traffic enlarging the denominator for
*both* engines, not from the VM paying extra.

Two conclusions:

- This is worth fixing on its own merits — a `u32::to_string()` and an
  `Arc<[u8]>` allocation per element read is indefensible on a hot path, and
  every allocation removed also shrinks GC pressure.
- It is **not** a prerequisite for #524's eligibility expansion, and no other
  work should be sequenced behind it. The measured per-entry cost in #539 is a
  much larger effect.

## Suggested approach

Port `eval_member`'s pre-`ToPropertyKey` numeric-index check into
`member_get_computed`, covering the typed-array and array-object cases it
already handles, keeping the slow path for every other key. Both sites should
call one shared helper rather than growing a second copy of the predicate.

`root_operand_stack`'s per-access cost is a separate question: the fast path
cannot reach a getter, a proxy trap, or a `ToPropertyKey` coercion, so it has no
nested safepoint to protect against and can skip the re-rooting. That should be
argued explicitly in the change rather than assumed.

## Validation

- test262 in both execution modes, no regression.
- A bytecode unit test asserting a typed-array element read through a compiled
  body matches the tree-walker for in-bounds, out-of-bounds, negative,
  fractional and `-0` indices, plus a detached buffer and a `Proxy` base.
- `bench_opmix.js`: `elem`'s absolute saving should rise above `arith`'s by more
  than noise. Given the bound above, expect a small effect — if it is
  unmeasurable, that is the honest outcome to record, not a reason to keep
  looking.

## References

- #526 (the audit; report and raw data under `docs/perf/2026-08-26/`)
- #539 (per-entry cost — the larger effect on the same code path)
- #524 (eligibility expansion — independent of this)

### PR DESCRIPTION
**Pull request by pmatos**
## Summary

- extract the tree-walker's typed-array/plain-array numeric-index read shortcut into one shared property-access helper
- let `Op::GetElement` use that helper before `ToPropertyKey`, avoiding the numeric string and `JsPropertyKey` allocations on fast-path hits
- skip `root_operand_stack` on hits because the helper cannot execute user code or reach an engine safepoint; pending operands remain rooted by `gc_bytecode_roots`
- preserve the existing slow path for proxies, shadowed array indices, holes, fractional/negative keys, numeric `-0`, and all other misses
- add bytecode/tree-walker parity coverage for the requested typed-array and array edge cases

The new tests are characterization-green before the optimization because the old slow path is already spec-correct; JavaScript-observable behavior cannot distinguish its internal allocations. I did not add an implementation-coupled allocation/call-count hook merely to manufacture a red test. The proposed sibling-GC test is similarly non-discriminating now that every VM push is independently rooted, so the change instead retains the existing GC/root-balance regressions and documents the no-safepoint argument at the carve-out.

## Validation

- `./scripts/lint.sh`
- `cargo build --release`
- `cargo test --release` — 625 passed, 1 ignored; integration and test262 smoke tests passed
- focused test262 areas (`TypedArrayConstructors/internals/Get`, property accessors, `TypedArray`, `Array`) — 9,091/9,091 passed in default mode and 9,091/9,091 with `--bytecode`
- full test262 default mode — 99,895 passed, 12 known failures, 0 regressions against `origin/main:test262-pass.txt`
- full test262 `--bytecode` mode — 99,895 passed, 12 known failures, 0 regressions against `origin/main:test262-pass.txt`

## Performance

Seven uninstrumented branch runs of `bench_opmix.js` produced median absolute savings of 0.294 µs/iteration for `arith` and 0.379 µs/iteration for `elem`, putting `elem` ahead by about 0.085 µs/iteration. An isolated same-host `origin/main` build was sampled as a baseline, but shared-host contention caused multi-fold timing swings and its median gap was larger, not smaller. The before/after effect is therefore honestly unmeasurable on this run; no performance claim beyond removing the known allocation path is made.

Closes #538

### CODE REVIEW

File: src/interpreter/property.rs [hunk 1]
```diff
@@ -123,6 +123,43 @@ impl Interpreter {
...
+            if index == trunc && !index.is_nan() && !index.is_sign_negative() {
+                return Some(JsValue::UNDEFINED);
+            }
+        }
+
+        if let Some(elems) = obj_borrow.array_elements() {
+            let trunc = index.trunc();
+            if index == trunc && index >= 0.0 && (index as usize) < elems.len() {
+                let idx = index as usize;
+                let key_str = (idx as u32).to_string();
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/interpreter/property.rs]: Fixed in `18078ef`.

Changed: `numeric_index_fast_get` now formats the shadow-descriptor lookup key into a 10-byte stack buffer (`format_u32_stack`) instead of `(idx as u32).to_string()`, so a successful dense-array numeric read (`a[i]`) no longer allocates.

Validation: `cargo test --bin jsse --release` (625 passed), `./scripts/lint.sh` (rustfmt + clippy incl. `perf-counters` feature), plus a manual check that a `defineProperty`-shadowed numeric index still overrides the fast path.
